TwitchChatVotingProxy/DiscordClient/DiscordBot.cs

```diff
@@ -0,0 +1,196 @@
+        {
+            string title = "Time for a new effect! Vote between:";
+            string optMsg = string.Empty;
+            string footer = string.Empty;
+            foreach (IVoteOption opt in options)
+            {
+                optMsg += $"{opt.Matches[0]} : {opt.Label} (Votes: {opt.Votes})\n";
+            }
+            if (votingMode == EVotingMode.PERCENTAGE)
+            {
+                footer = "Votes will affect the chance for one of the effects to occur.";
+            }
+
+            var embed = new EmbedBuilder
+            {
+                Title = title,
+                Description = optMsg,
+            };
+            if (!string.IsNullOrEmpty(footer))
+            {
+                embed.WithFooter(footer);
+            }
+
+            embed.WithAuthor(_client.CurrentUser);
+            embed.WithCurrentTimestamp();
+
+            curMsg.ModifyAsync(msg => msg.Embed = embed.Build());
+        }
+
+        public async Task ButtonHandler(SocketMessageComponent component)
+        {
+            foreach (var item in buttonIds)
+            {
+                if (item.Value == component.Data.CustomId)
+                {
+                    OnVoteReceived?.Invoke(this, new BotVoteEventArgs()
+                        {
+                            UserId = component.User.Id,
+                            Vote = item.Key
+                        }
+                    );
+                    
+                    await component.RespondAsync($"You voted for effect #{item.Key}", ephemeral: true);
+                    break;
+                }
+            }
+        }
+
+        private Task Hearbeat(int a, int b)
+        {
+            ConnectionState nowState = _client.ConnectionState;
+            if (nowState != connectionState)
+            {
+                connectionState = nowState;
+                switch (connectionState)
+                {
+                    case ConnectionState.Connected:
+                        OnBotConnected?.Invoke(this, EventArgs.Empty);
+                        break;
+                    case ConnectionState.Disconnected:
+                        OnBotDisconnected?.Invoke(this, EventArgs.Empty);
+                        break;
+                }
+
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void StartBot()
+        {
+            logger.Information("Starting bot!");
+            _client = new DiscordSocketClient();
+
+            _client.LoginAsync(TokenType.Bot, credentials.OAuth);
+            _client.StartAsync();
+            _client.SetGameAsync("Chaos Mod", type: ActivityType.Playing);
+
+            _client.LatencyUpdated += Hearbeat;
+            _client.ButtonExecuted += ButtonHandler;
+        }
+
+        public void StopBot()
+        {
+            logger.Information("Stopping bot!");
+            _client.LatencyUpdated -= Hearbeat;

```

Comment:
With that you won’t get the disconnected handling, if you remove the handler before disconnecting